source/plugins/ruby/kubelet_utils.rb

```diff
@@ -41,6 +41,108 @@ def get_node_capacity
       end
     end
 
+    def get_node_allocatable(winNode)
+      begin
+        cpu_capacity = 1.0
+        memory_capacity = 1.0
+        cpu_allocatable = 1.0
+        memory_allocatable = 1.0
+        capacity_response = CAdvisorMetricsAPIClient.getAllMetricsCAdvisor
+        if !capacity_response.nil? && !capacity_response.body.nil?
+          all_metrics = capacity_response.body.split("\n")
+          #cadvisor machine metrics can exist with (>=1.19) or without dimensions (<1.19)
+          #so just checking startswith of metric name would be good enough to pick the metric value from exposition format
+          cpu_capacity = all_metrics.select { |m| m.start_with?("machine_cpu_cores") }.first.split.last.to_f * 1000
+          @log.info "get_node_allocatable::CPU Capacity #{cpu_capacity}"
+          memory_capacity_e = all_metrics.select { |m| m.start_with?("machine_memory_bytes") }.first.split.last
+          memory_capacity = BigDecimal(memory_capacity_e).to_f
+          @log.info "get_node_allocatable::Memory Capacity #{memory_capacity}"
+        end
+
+        if !winNode.nil?
+          winNode = nil
+        end
+
+        allocatable_response = CAdvisorMetricsAPIClient.getCongifzCAdvisor(winNode)
+
+        begin
+          kubereserved_cpu = JSON.parse(allocatable_response.body)["kubeletconfig"]["kubeReserved"]["cpu"]
+          if kubereserved_cpu.nil? || kubereserved_cpu == ""
+            kubereserved_cpu = "0"
+          end
+          @log.info "get_node_allocatable::kubereserved_cpu  #{kubereserved_cpu}"
+        rescue => errorStr
+          @log.error "Error in get_node_allocatable::kubereserved_cpu: #{errorStr}"
+          kubereserved_cpu = "0"
+          ApplicationInsightsUtility.sendExceptionTelemetry("Error in get_node_allocatable::kubereserved_cpu: #{errorStr}")
+        end 
+
+        begin
+          kubereserved_memory = JSON.parse(allocatable_response.body)["kubeletconfig"]["kubeReserved"]["memory"]
+          if kubereserved_memory.nil? || kubereserved_memory == ""
+            kubereserved_memory = "0"
+          end
+          @log.info "get_node_allocatable::kubereserved_memory #{kubereserved_memory}"
+        rescue => errorStr
+          @log.error "Error in get_node_allocatable::kubereserved_memory: #{errorStr}"
+          kubereserved_memory = "0"
+          ApplicationInsightsUtility.sendExceptionTelemetry("Error in get_node_allocatable::kubereserved_cpu: #{errorStr}")
+        end 
+        begin
+          systemReserved_cpu = JSON.parse(allocatable_response.body)["kubeletconfig"]["systemReserved"]["cpu"]
```

Comment:
similarly here JSON.parse. doing one time JSON.parse and reusing the parsed object would help on perf side.

Reply:
updated